### PR TITLE
Adds a format checker to the workflows, which fails if the code has not been formatted, adds a dev folder with instructions to setup a nightly env + point the git hooks to the local folder

### LINKED ISSUE #1085: Add `cargo fmt --check` to CI

### PR DESCRIPTION
**Pull request by Haaroon**
### What changes were proposed in this pull request?

Adds a format checker to the workflows, which fails if the code has not been formatted
- checks out nightly cargo in a workflow
- runs the format check
- add a new `dev` folder with a bootstrap.sh file that setups hooks and the dev environment 

### Why are the changes needed?

prevent dirty formatted code

### Does this PR introduce any user-facing change? If yes is this documented?

no

### How was this patch tested?

both by hand and in the workflow 

### Issues

resolves #1085 

### Are there any further changes required?

no

### CODE REVIEW

**Comment by Haaroon** [comment]
yeah sorry i wrote the wrong commit message 🙃

**Comment by Haaroon** [comment]
this run should fail because i did not format the code.

**Comment by Haaroon** [comment]
ready

File: dev/bootstrap.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  echo "$(basename ${BASH_SOURCE[0]}): $@"
+}
+
+install_hooks() {
+  git config core.hooksPath \
+    || git config core.hooksPath ./dev/hooks
+}
+
+install_rust_nightly() {
+  rustup toolchain install nightly
```
Comment (fabianmurariu) [inline dev/bootstrap.sh:17]: why are we installing nightly?

File: dev/hooks/pre-commit [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+# Run rustfmt on the whole repository using nightly toolchain
+cargo +nightly fmt --all -- --check
```
Comment (fabianmurariu) [inline dev/hooks/pre-commit:4]: why nightly?

File: dev/bootstrap.sh (hunk 1 above)
Comment (Haaroon) [reply dev/bootstrap.sh:17]: so that uses can run the formatter, otherwise it will fail to run in the hook.

File: dev/hooks/pre-commit (hunk 2 above)
Comment (Haaroon) [reply dev/hooks/pre-commit:4]: because it has the features for formatting imports which stable does not have

**Pull request merged**